### DISCUSSION TITLE
OpenGL Backend Support (missing features)

### DISCUSSION THREAD

**Discussion by shingen75** [alandefreitas/matplotplusplus#380] [Q&A]
Hi I am doing a simple compare between the GnuPlot backend and the OpenGL backend. I was looking into making a contour plot or a 3D plot but before all of that I was trying to see what is missing between a simple 2D plot 

The following code 

int main() {
    using namespace matplot;
    std::vector<double> x = linspace(0, 2 * pi);
    std::vector<double> y = transform(x, [](auto x) { return sin(x); });

    plot(x, y, "-o");
    hold(on);
    plot(x, transform(y, [](auto y) { return -y; }), "--xr");
    plot(x, transform(x, [](auto x) { return x / pi - 1.; }), "-:gs");
    plot({ 1.0, 0.7, 0.4, 0.0, -0.4, -0.7, -1 }, "k");

    show();
    return 0;
}

produces the following plot 
![image](https://github.com/alandefreitas/matplotplusplus/assets/144074460/bbdc000d-64e3-4e0b-9adf-b3c66d44b4e1)
gnuplot output 

Now for the OpenGL backend I have changed the code to the following 
    using namespace matplot;

    auto f = figure<backend::opengl>(true);
    auto ax = f->current_axes();

    using namespace matplot;
    std::vector<double> x = linspace(0, 2 * pi);
    std::vector<double> y = transform(x, [](auto x) { return sin(x); });

    ax->plot(x, y, "-o");
    ax->hold(on);
    ax->plot(x, transform(y, [](auto y) { return -y; }), "--xr");
    ax->plot(x, transform(x, [](auto x) { return x / pi - 1.; }), "-:gs");
    ax->plot({ 1.0, 0.7, 0.4, 0.0, -0.4, -0.7, -1 }, "k");

    // Start rendering
    while (!f->should_close()) 
    {
        f->draw();
    }

which produces the following 
![image](https://github.com/alandefreitas/matplotplusplus/assets/144074460/db970053-408f-44c1-8438-07993fac06f9)

I am not sure if the above code is missing anything but I wrote based on the test example for the OpenGL backend. 

Looking at the plots, the OpenGL backend is lacking the following: 
1. No axis information 
2. The line patterns are not supported
3.  Coordinate information is not supported 

I assume for the above that OpenGL is still very early and likely 3D and contour plots are not supported. Is this correct ?
I understand the difficulty of providing the capability but I was wondering if there is a timeline OpenGL backend would have additional feature support. 

Appreciate the great effort put into this work.

**Comment by alandefreitas** [comment]
Yes. It's missing lots of features. It has always been experimental and never left this state. The idea is that gnuplot and opengl would be two of many backends. Other backends could be the console, QT, ImGUI, etc... The real problem is I never had the time to work on that outside my day job. I never found anyone to work on that either.

If I were doing that today (I probably won't have the time any time soon), I think I would refactor the backend interface to fix problems we have identified along the way. In the end, this would mean matplot++ v2.0.0. 

- Backends would only accept primitives: lines, colors, etc... 
- Points and other primitives would be allowed to have more generic formats (Boost.GIL is a good example of how these primitives can be generalized) to accommodate backend requirements without transforms. 
- `figure` would look more like a vector of type erased objects so the interdependence with specific plot types would be gone and it would be easier for people to come up with their own plot types. 
- There would be no default registry of figure objects, "classic" mode (plotting without an explicit figure), or non-quiet mode.

This is kind of similar to the python matplotlib backend interface, although types make it more complex. These changes mean gnuplot would be discontinued unless we create a gnuplot backend that can draw things in terms of primitive values (which would also make it gnuplot kind of useless in any way that's not for backward compatibility). The first backends I would implement in this new model would be SDL and console, although there's room to discuss alternatives.
